torchrecipes/paved_path/charnn/main.py

```diff
@@ -0,0 +1,187 @@
+import socket
+import uuid
+from typing import Optional, Tuple
+
+import hydra
+import torch
+import torch.distributed as dist
+from omegaconf import DictConfig
+from torch.nn.parallel import DistributedDataParallel
+from torch.utils.data import random_split
+
+from char_dataset import CharDataset, get_dataset
+from model import GPT, GPTConfig, OptimizerConfig
+from trainer import Checkpoint, Trainer, TrainerConfig, load_checkpoint
+from utils import get_realpath, sample
+
+logger = logging.getLogger(__name__)
+
+
+def get_fq_hostname() -> str:
+    return socket.getfqdn(socket.gethostname())
+
+
+def set_env() -> None:
+    os.environ["RANK"] = os.environ.get("RANK", "0")
+    os.environ["WORLD_SIZE"] = os.environ.get("WORLD_SIZE", "1")
+    os.environ["MASTER_PORT"] = os.environ.get("MASTER_PORT", "29830")
+    os.environ["MASTER_ADDR"] = os.environ.get("MASTER_ADDR", "localhost")
+    os.environ["LOCAL_RANK"] = os.environ.get("LOCAL_RANK", "0")
+    os.environ["TORCHELASTIC_RUN_ID"] = os.environ.get(
+        "TORCHELASTIC_RUN_ID", str(uuid.uuid4()).split("-")[0]
+    )
+
+
+def get_job_name() -> str:
+    uid = os.environ["TORCHELASTIC_RUN_ID"]
+    return f"test-job-{uid}"
+
+
+def get_device() -> torch.device:
+    if  torch.cuda.is_available():
+        local_rank = int(os.environ.get("LOCAL_RANK", "0"))
+        device = torch.device(f"cuda:{local_rank}")
+        torch.cuda.set_device(device)
+    else:
+        device = torch.device("cpu")
+    return device
+    
+
+
+def get_ddp_model_and_optimizer(
+    gpt_config: GPTConfig, opt_config: OptimizerConfig, checkpoint: Optional[Checkpoint]
+) -> Tuple[torch.nn.Module, torch.optim.Optimizer]:
+    # Create new GPT Model on CPU
+    model = GPT(gpt_config)
+    # Load GPT model from checkpoint if present
+    if checkpoint:
+        model.load_state_dict(checkpoint.model_state)
+    device = get_device()
+    device_ids = None
+    if device.type == "cuda":
+        model = model.to(device)
+        device_ids = [device.index]
+    model = DistributedDataParallel(
+        model,
+        device_ids=device_ids,
+    )
+    optimizer = torch.optim.AdamW(
+        model.parameters(), lr=opt_config.lr, weight_decay=opt_config.weight_decay
+    )
+    return model, optimizer
+
+
+def get_model_and_optimizer(
+    type: str,
+    gpt_config: GPTConfig,
+    opt_config: OptimizerConfig,
+    checkpoint: Optional[Checkpoint],
+) -> Tuple[torch.nn.Module, torch.optim.Optimizer]:
+    if type == "ddp":
+        return get_ddp_model_and_optimizer(gpt_config, opt_config, checkpoint)
+    
+    raise RuntimeError(f"Unknown type: {type}. Allowed values: [ddp]")
+
+
+def setup_process_group() -> None:
+    device = get_device()
+    rank = int(os.environ["RANK"])
+    world_size = int(os.environ["WORLD_SIZE"])
+    if device.type == "cuda":
+        dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    else:
+        dist.init_process_group("gloo", rank=rank, world_size=world_size)
```

Comment:
assuming we're launching with torchrun/torchx, this would initialize both device & process group: 

https://github.com/pytorch/tnt/blob/5fd77ae9ad8357079cebd14532cb40a2ce8e5761/torchtnt/utils/env.py#L34-L38